grafana/grafana
# Query history: Cleanup

**What this PR does / why we need it**:
In this PR, we are creating cleanup actions - **delete every unstarred query older than 14 days** and we are adding **removing of stars when the user is delete from the org**. 

**Which issue(s) this PR fixes**:
Part of https://github.com/grafana/grafana/issues/44327

**How to test:**
To test the functionality you can add this snippet to https://github.com/grafana/grafana/blob/main/public/app/features/explore/state/query.ts#L317  and you need to `import { getBackendSrv } from '@grafana/runtime';`

You also have to enable query history `enabled = true` flag in custom.ini. 

```
  //To test adding of query history
  const dataSourceUid = queries[0]?.datasource?.uid;
  if (dataSourceUid) {
    getBackendSrv().post(`/api/query-history`, {
      dataSourceUid,
      queries,
    });
  }
  
  if (Math.random() * 10 < 5) {
    getBackendSrv().post(`/api/query-history/star/${result.uid}`);
  }
  // End of test
```

1. To test deleting of queries, you need to change `queryHistoryRetention := time.Minute * 2` and `ticker := time.NewTicker(time.Minute * 1)` in in `pkg/services/cleanup/cleanup.go` to see if queries older than 2 minutes were removed. 

2. After this, to test unstarring of queries, you need to create a new user, run some queries in Explore (some of them will be starred), and then remove the user. When the user is removed, all their starred queries should be removed.
